DataDog/dd-trace-rb
# Add info transport endpoint

**What does this PR do?**

Implement the `/info` endpoint without impacting the current trace endpoint.

**Motivation**

The goal of this commit is twofold:

- prepare for endpoint and configuration negotiation with the agent
- lay ground for reaching to other endpoints than the trace API

To that end, it:

- duplicates and alters a few files, noting the issues encountered towards full reusability of the current transport implementation.
- implements requesting to the `GET /info` endpoint
- does not implement processing the response nor any form of actual negotiation using that response

**Additional Notes**

This may be easier to review by comparing between `lib/ddtrace` and `lib/datadog/core`, e.g:

```
diff --color -w -u lib/{ddtrace,datadog/core}/transport/http/builder.rb
```

The short-term goal is to enable implementing additional endpoints, such as `/v0.7/config` for remote configuration.

Many comments have been dropped in the code itself to act as persistent analysis towards a possible later longer-term refactoring.

**How to test the change?**

The code is isolated and uncalled, so CI should stay green.

Manual testing:

- Step 1: start the agent
```
docker run --rm -it -e DD_APM_ENABLED=true -e DD_ENV=${YOUR_ENV} -e DD_HOSTNAME=${YOUR_HOSTNAME} -e DD_BIND_HOST=0.0.0.0 -e DD_APM_NON_LOCAL_TRAFFIC=true -e DD_API_KEY=${YOUR_API_KEY}-e DD_REMOTE_CONFIGURATION_ENABLED=true -p 8126:8126 datadog/agent:7.42.0
```
- Step 2: hit it with the following code

```
require 'ddtrace'

Datadog.configure do |c|
    c.agent.host = "#{YOUR_AGENT_HOST}"
    # c.diagnostics.debug = true
end

transport_options = {
  agent_settings: Datadog::Core::Configuration::AgentSettingsResolver.call(Datadog.configuration)
}

require 'datadog/core/transport/http'

transport_root = Datadog::Core::Transport::HTTP.root(**transport_options.dup)

puts transport_root.send_info.inspect
```
